isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp

```diff
@@ -76,6 +86,3886 @@ namespace isobus
+
+		/// @brief Removes an object reference from another object. All fields must exactly match for the object to be removed.
+		/// This is because objects can have multiple of the same child at different places, so we can't infer which one to
+		/// remove without the exact position.
+		/// @param[in] objectIDToRemove The object ID of the child to remove
+		/// @param[in] relativeXLocation The X offset of this object to its parent
+		/// @param[in] relativeYLocation The Y offset of this object to its parent
+		void remove_child(std::uint16_t objectIDToRemove, std::int16_t relativeXLocation, std::int16_t relativeYLocation);
+
+		/// @brief Removes the last added child object.
+		/// This is meant to be a faster way to deal with objects that only have a max of 1 child.
+		void pop_child();
+
+	protected:
+		/// @brief Storage for child object data
+		class ChildObjectData
+		{
+		public:
+			/// @brief Default constructor for child object data with default values
+			ChildObjectData() = default;
+
+			/// @brief Constructor that initializes all members with parameters
+			/// @param[in] objectId The object ID of this child object
+			/// @param[in] x The x location of this child relative to the parent object
+			/// @param[in] y The y location of this child relative to the parent object
+			ChildObjectData(std::uint16_t objectId,
+			                std::int16_t x,
+			                std::int16_t y);
+			std::uint16_t id = NULL_OBJECT_ID; ///< Object identifier. Shall be unique within the object pool.
+			std::int16_t xLocation = 0; ///< Relative X location of the top left corner of the object
+			std::int16_t yLocation = 0; ///< Relative Y location of the top left corner of the object
+		};
+
+		/// @brief A helper function to swap a child object of specified type with a new object.
+		/// This is meant to be used to replace attributes of which there can be only 1, such as a string variable in an output string.
+		/// @param[in] newID The new child object's ID
+		/// @param[in] typeToRemove The object type of both the current child, and the new child (they must match)
+		/// @returns True if the replacement succeeded, otherwise false
+		bool replace_only_child_of_type(std::uint16_t newID, VirtualTerminalObjectType typeToRemove);
+
+		/// @brief A helper function to get the first child object of the specified type
+		/// @param[in] typeToFind The object type to find
+		/// @returns The object ID of the first child object of the specified type, or NULL_OBJECT_ID if none were found
+		std::uint16_t get_first_child_of_type(VirtualTerminalObjectType typeToFind) const;
+
+		VTColourTable &colourTable; ///< A reference to the current object pool's colour table. Useful for rendering most objects.
+		std::map<std::uint16_t, std::shared_ptr<VTObject>> &thisObjectPool; ///< A pointer to the rest of the object pool. Convenient for lookups by object ID.
```

Comment:
That's fair - the way I'm consuming these classes myself is safe I think but that doesn't mean it would be for others. Might be better to store a weak pointer or something

Reply:
Or maybe just require an argument in the functions that use it? It doesn't seem to be used that often IIRC

Reply:
It's mostly there for `std::shared_ptr<VTObject> get_object_by_id(std::uint16_t objectID) const;` In my VT server I call this hundreds of times haha. My GUI objects actually inherit from each of these objects, and then also inherit from the GUI framework, so inside their `paint` routine I can get objects from the parent object pool easily.

If we could force people to interact with creating these objects only through some other class, then we could ensure the reference is valid... like, we could make the constructors badged, or private with a `friend` class of the object pool or something. IDK 🤔

Reply:
Ohhh I see! I looked a bit in the VT server, and noticed in almost all places where you call `get_object_by_id` you also have a `VTServerManagerWorkingSet` to your disposal. Maybe it would be an idea to move this function to there? Since that object should also have access to the object pool. 

Also theoretically, one might want to handle objects independent from the complete pool. I think that is a little difficult right now by requiring an object pool for an object. 

> If we could force people to interact with creating these objects only through some other class, then we could ensure the reference is valid... like, we could make the constructors badged, or private with a friend class of the object pool or something.

I think instead of limiting the interaction with the objects, the approach above, where specific operations can only happen with some kind of middle-man that holds the reference as value/smart_pointer instead, might be a good tradeoff?

Reply:
> If we could force people to interact with creating these objects only through some other class, then we could ensure the reference is valid

I mean, that would only hold if the other class holds the pool as a value **and** lives longer than the vt-object. The last one you can't really guarantee without an extra dependency I think

Reply:
Removed stored references to both the object tree and colour table.